url4c41254592_topskychen_AuthTopk_tgz-pJ8-quickhull3d_QuickHull3DTestJ8/src/utility/geo/buildIndexDL.java

```diff
@@ -0,0 +1,186 @@
+/**
+ * 
+ */
+package utility.geo;
+
+import java.io.BufferedOutputStream;
+import java.io.DataOutputStream;
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
+import java.io.FileReader;
+import java.io.IOException;
+import java.io.LineNumberReader;
+import java.util.ArrayList;
+import java.util.Scanner;
+
+import mesh.PDelaunay;
+import utility.Compare.DataOfPoint;
+import utility.security.Point;
+import utility.security.RSA;
+
+/**
+ * @author chenqian
+ * Build index for data of line. The index has two files. 
+ * One is *.dat, and the other is *.idx, which are data and index respectively.
+ * We dont use btree this time. We use a hash map store the index, and every 
+ * tuple has <position, length>.
+ * The input data should be formatted to [0, 1].
+ */
+public class buildIndexDL {
+
+	
+	public static ArrayList<long[]> points = null;	
+	public static int ThreadNum = 8;
+	public static boolean[] threadStatus = new boolean[ThreadNum];
+	public static RSA rsa = new RSA(); // sign for every data of point.
+
+	/**
+	 * Load data.
+	 * 
+	 * */
+	public static void loadData(String sourceFileName){
+		points = new ArrayList<long[]>();
+		
+		try {
+			String line = null;
+			LineNumberReader lr = new LineNumberReader(new FileReader(sourceFileName));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *lr*. Execution paths that do not contain closure statements were detected. Ensure *lr* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).